src/Chapter3.hs

```diff
@@ -1006,6 +1161,79 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data MonsterActions = MAttack | Run
+    deriving (Show, Eq, Enum, Bounded)
+
+data Monster' = M 
+    { mHealth  :: Health
+    , mAttack  :: Attack
+    , mActions :: MonsterActions
+    } deriving Show
+
+data KnightActions 
+    = KAttack
+    | Heal
+    | SpellCast
+    deriving (Show, Eq, Enum, Bounded)
+  
+data Knight' = K
+    { kHealth  :: Health
+    , kAttack  :: Attack
+    , kDefense :: Defense
+    , kActions :: KnightActions
+    } deriving Show
+
+cycleMActions :: MonsterActions -> MonsterActions
+cycleMActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
+
+cycleKActions :: KnightActions -> KnightActions
+cycleKActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
+
+getMonsterDamage :: Health -> Attack -> Health
+getMonsterDamage (Health h) (Attack a) = Health (h - a)
+
+getKnightDamange :: Health -> Defense -> Attack -> Health
+getKnightDamange (Health h) (Defense d) (Attack a) = Health ((h + d) - a)
+
+skipKAttack :: KnightActions -> Attack -> Attack
+skipKAttack KAttack a = a
+skipKAttack _ _ = Attack 0
+    
+skipMAttack :: MonsterActions -> Attack -> Attack
+skipMAttack MAttack a = a
+skipMAttack _ _ = Attack 0
+
+class Battle a b where {
+    battle :: a -> b -> String
+}
+
+instance Battle Knight' Monster' where
+    battle :: Knight' -> Monster' -> String
```

Comment:
I see that you use multiparam typeclasses to model this scenario. This is fine, but this particular solution involves writing a lot of boilerpate. For example, to be able to have  all possible combinations, you need to implement an instance for each pair. E.g:

```haskell
instance Battle Monster Knight
instance Battle Knight Knight
instance Battle Monster Monster
```

You can imagine that this will involve quite a lot of boilerplate and code duplication :disappointed: 

One way to solve the problem is to implement a typeclass like this:

```haskell
class Fighter f where
    fight :: Fighter other => f -> other -> (other, f)
```

The `fight` function, when you implement instance, knows all implementation details of the `f` type, but at the same time it allows to work with the value of type `other` in a generic and polymorphic way. This function represents a single round of the fight. With this approach you only need to implement a single instance for each type:

```haskell
instance Fighter Monster
instance Fighter Knight
```

and you will be able to implement a general recursive fight function.

To implement instances, you probably need to add more methods to the `Fighter` typeclass. Something like `attack :: f -> Attack` and `isAlive :: f -> Bool` but these is just an idea :slightly_smiling_face: